### PR TITLE
React JSX Highmaps

### PR DESCRIPTION
**Pull request by whawker**
For #56. 

Just looking for any feedback, if you don't mind @anajavi ?

### CODE REVIEW

**Comment by anajavi** [comment]
Although I'm not familiar with Highmaps, it looks very nice indeed!

I try to check it properly later, when I have the time. Now I did do only a cursory scan.

The MapLoader component left me thinking: is data loading this library's concern?
- If the data is behind authentication, the MapLoader needs to have more than url in props
- Same for POST requests and headers like Accept: application/json
- It needs fetch-polyfill to work on IE11

Would it be better to move the MapLoader to the examples? That way everybody could customise it the way they like.

Anyway, awesome work!

**Comment by whawker** [comment]
Hmm yeah, fair point about the MapLoader, I think you're right in hindsight. I think I'll link to an example implementation in the readme.

I also think I'll change the default landing readme for this project, and just link out to separate packages instead.

File: packages/react-jsx-highmaps/.babelrc [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  ],
+  "env": {
+    "test": {
+      "plugins": ["@babel/transform-modules-commonjs"]
+    },
+    "production": {
+      "plugins": ["transform-react-remove-prop-types",
+        ["@babel/transform-runtime", { "useESModules": true }]
+      ]
+    }
```
Comment (anajavi) [inline packages/react-jsx-highmaps/.babelrc]: env:es is missing:
```
    "es": {
      "plugins": ["./build/use-lodash-es",
        "transform-react-remove-prop-types",
        ["@babel/transform-runtime", { "useESModules": true }]
      ]
    }
```

File: packages/react-jsx-highmaps/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+We aim to pass all configuration options using the same name, so we use [Highcharts' documentation](http://api.highcharts.com/highcharts) to figure out how to achieve the same with React JSX Highcharts.
+
+### Note:
+
+There are **two** exceptions to the above;
+
+#### Exception 1
+
+Where Highcharts **events** are concerned - instead of passing `events` as an object, we use the React convention *onEventName*.
+
```
Comment (anajavi) [inline packages/react-jsx-highmaps/README.md]: Here an example of events would clarify what needs to be done.

File: packages/react-jsx-highmaps/build/use-lodash-es.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module.exports = function() {
+  return {
+    visitor: {
+      ImportDeclaration(path) {
+        var source = path.node.source
+        source.value = source.value.replace(/^lodash($|\/)/, 'lodash-es$1')
+      },
+    },
+  }
+}
```
Comment (anajavi) [inline packages/react-jsx-highmaps/build/use-lodash-es.js]: As there are no commonjs modules published in any package, we could get rid of this and lodash dependency by just using lodash-es in source code. That would simplify things.

File: packages/react-jsx-highmaps/src/index.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+export * from 'react-jsx-highcharts';
+
+// Charts
+export { default as HighchartsMapChart } from './components/HighchartsMapChart';
+
+// Graph Parts
+export { default as MapNavigation } from './components/MapNavigation';
+export { default as XAxis } from './components/XAxis';
+export { default as YAxis } from './components/YAxis';
```
Comment (anajavi) [inline packages/react-jsx-highmaps/src/index.js]: XAxis and YAxis are already exported by:
```javascript
export * from 'react-jsx-highcharts';
```
Might be better to name these as MapXAxis and MapYAxis, so they don't get confused with the highcharts ones. It seems that they behave differently than the ones from react-jsx-highcharts, but I haven't tested.

File: packages/react-jsx-highmaps/webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        loader: 'babel-loader',
+        exclude: /node_modules(\\|\/)(?!react-jsx-highcharts)/
+      }
+    ]
+  },
+
+  plugins: [
+    new webpack.DefinePlugin({
+      'process.env': {
+        'NODE_ENV': JSON.stringify(process.env.NODE_ENV)
```
Comment (anajavi) [inline packages/react-jsx-highmaps/webpack.config.js]: mode = 'production' sets this also: https://webpack.js.org/concepts/mode/

File: packages/react-jsx-highmaps/build/use-lodash-es.js (hunk 3 above)
Comment (anajavi) [reply packages/react-jsx-highmaps/build/use-lodash-es.js]: It could be changed in all packages.

File: examples/Map/App.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+const labelFormatter = function () {
+  if (this.point.value) {
+    return this.point.name
+  }
+}
+
+const App = () => (
+  <div className="app">
+    <Fetch url="https://code.highcharts.com/mapdata/custom/europe.geo.json">
```
Comment (anajavi) [inline examples/Map/App.js]: This is nice! Now there's no need to explain fetch polyfilling in the readme :)

File: packages/react-jsx-highmaps/src/index.js (hunk 4 above)
Comment (whawker) [reply packages/react-jsx-highmaps/src/index.js]: I think I might remove `export * from 'react-jsx-highcharts';` and export only the parts that are relevant to maps. Unlike Highstock, I don't think people should be trying to build charts with this package.

File: packages/react-jsx-highmaps/build/use-lodash-es.js (hunk 3 above)
Comment (whawker) [reply packages/react-jsx-highmaps/build/use-lodash-es.js]: Would this mean we could also remove the need for lodash-webpack-plugin?

File: packages/react-jsx-highmaps/src/components/HighchartsMapChart/HighchartsMapChart.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    const geojson = this.getGeoJSON(map);
+
+    return (
+      <BaseChart
+        chart={{ ...chart, map: geojson }}
+        mapNavigation={{ enabled: false }}
+        xAxis={{ id: 'xAxis' }}
+        yAxis={{ id: 'yAxis' }}
+        {...rest}
+        callback={this.callback(geojson)}
```
[the remark this answers is not in the document]
Comment (whawker) [reply packages/react-jsx-highmaps/src/components/HighchartsMapChart/HighchartsMapChart.js]: I believe [this modification](https://github.com/whawker/react-jsx-highcharts/pull/163/files#diff-f084ff751cb53f3465484c946894fb4bR20) should fix this problem, as `this.getGeoJSON` always returns the same object reference?

File: packages/react-jsx-highmaps/src/index.js (hunk 4 above)
Comment (anajavi) [reply packages/react-jsx-highmaps/src/index.js]: That might be better. Could also be:
```javascript
export { withHighcharts as withHighmaps } from 'react-jsx-highcharts';
```

File: packages/react-jsx-highmaps/build/use-lodash-es.js (hunk 3 above)
Comment (anajavi) [reply packages/react-jsx-highmaps/build/use-lodash-es.js]: babel-plugin-lodash could be then removed. 

lodash-webpack-plugin removes some inner parts from lodash itself, so it is still useful.

File: packages/react-jsx-highmaps/src/components/HighchartsMapChart/HighchartsMapChart.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (anajavi) [reply packages/react-jsx-highmaps/src/components/HighchartsMapChart/HighchartsMapChart.js]: Looks like it fixes the problem.

File: packages/react-jsx-highmaps/build/use-lodash-es.js (hunk 3 above)
Comment (anajavi) [reply packages/react-jsx-highmaps/build/use-lodash-es.js]: I can do it, if you don't mind?

Comment (whawker) [reply packages/react-jsx-highmaps/build/use-lodash-es.js]: No I don't mind at all, thanks!

**Comment by whawker** [comment]
Thanks very much for your help on this @anajavi, I'll publish some beta versions so we can check everything works as expect.

**Comment by anajavi** [comment]
Don't you have to publish react-jsx-highcharts also as this PR had changes to it too?

**Comment by whawker** [comment]
Yeah, that's the plan. Haven't published any betas for the Babel 7 stuff yet, so would be publishing those anyway.

**Pull request merged**